pkg/rpc/rpc.go

```diff
@@ -0,0 +1,135 @@
+package rpc
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/icinga/icinga-notifications/pkg/plugin"
+	"go.uber.org/zap"
+	"io"
+	"sync"
+)
+
+type Error struct {
+	cause error
+}
+
+type RPC struct {
+	writer          io.Closer // use encoder for writing instead
+	encoder         *json.Encoder
+	decoder         *json.Decoder
+	pendingRequests map[uint64]chan plugin.JsonRpcResponse
+	logger          *zap.SugaredLogger
+	lastRequestId   uint64
+
+	mu         sync.Mutex
+	errChannel chan error
+	rpcErr     *Error
+}
+
+func NewRPC(writer io.WriteCloser, reader io.Reader, logger *zap.SugaredLogger) *RPC {
+	rpc := &RPC{
+		writer:          writer,
+		encoder:         json.NewEncoder(writer),
+		decoder:         json.NewDecoder(reader),
+		pendingRequests: map[uint64]chan plugin.JsonRpcResponse{},
+		logger:          logger,
+		errChannel:      make(chan error, 1),
+	}
+
+	go rpc.processResponses()
+
+	return rpc
+}
+
+func (r *RPC) Call(method string, params json.RawMessage) (json.RawMessage, error) {
+	if err := r.getErr(); err != nil {
+		return nil, err
+	}
+
+	promise := make(chan plugin.JsonRpcResponse, 1)
+
+	r.mu.Lock()
+	r.lastRequestId++
+	newId := r.lastRequestId
+	r.pendingRequests[newId] = promise
+	err := r.encoder.Encode(plugin.JsonRpcRequest{Method: method, Params: params, Id: newId})
+	r.mu.Unlock()
+	if err != nil {
+		err = fmt.Errorf("failed to write request: %w", err)
+		r.setErr(err)
+
+		return nil, err
+	}
+
+	select {
+	case response := <-promise:
+		if response.Error != "" {
+			return nil, fmt.Errorf("plugin response contains error: %s", response.Error)
```

Comment:
I'd just use `errors.New(response.Error)` here, i.e. not prefix with an extra message. Rationale: that's the closest to what the corresponding function on the remote side returned, so `Call()` returns either what that function returned, or the special `rpc.Error`.